protocols/gossipsub/src/behaviour.rs

```diff
@@ -0,0 +1,1176 @@
+                    }
+
+                    // if the mesh needs peers add the peer to the mesh
+                    if let Some(peers) = self.mesh.get_mut(&subscription.topic_hash) {
+                        if peers.len() < self.config.mesh_n_low {
+                            debug!(
+                                "SUBSCRIPTION: Adding peer {:?} to the mesh",
+                                propagation_source,
+                            );
+                        }
+                        peers.push(propagation_source.clone());
+                    }
+                    // generates a subscription event to be polled
+                    self.events.push_back(NetworkBehaviourAction::GenerateEvent(
+                        GossipsubEvent::Subscribed {
+                            peer_id: propagation_source.clone(),
+                            topic: subscription.topic_hash.clone(),
+                        },
+                    ));
+                }
+                GossipsubSubscriptionAction::Unsubscribe => {
+                    if let Some(pos) = peer_list.iter().position(|p| p == propagation_source) {
+                        info!(
+                            "SUBSCRIPTION: Removing gossip peer: {:?} from topic: {:?}",
+                            propagation_source, subscription.topic_hash
+                        );
+                        peer_list.remove(pos);
+                    }
+                    // remove topic from the peer_topics mapping
+                    if let Some(pos) = subscribed_topics
+                        .iter()
+                        .position(|t| t == &subscription.topic_hash)
+                    {
+                        subscribed_topics.remove(pos);
+                    }
+                    // remove the peer from the mesh if it exists
+                    if let Some(peers) = self.mesh.get_mut(&subscription.topic_hash) {
+                        peers.retain(|peer| peer != propagation_source);
+                    }
+
+                    // generate a subscription even to be polled
+                    self.events.push_back(NetworkBehaviourAction::GenerateEvent(
+                        GossipsubEvent::Unsubscribed {
+                            peer_id: propagation_source.clone(),
+                            topic: subscription.topic_hash.clone(),
+                        },
+                    ));
+                }
+            }
+        }
+        trace!(
+            "Completed handling subscriptions from source: {:?}",
+            propagation_source
+        );
+    }
+
+    /// Heartbeat function which shifts the memcache and updates the mesh
+    fn heartbeat(&mut self) {
+        debug!("Starting heartbeat");
+
+        //TODO: Clean up any state from last heartbeat.
+
+        let mut to_graft = HashMap::new();
+        let mut to_prune = HashMap::new();
+
+        // maintain the mesh for each topic
+        for (topic_hash, peers) in self.mesh.clone().iter_mut() {
```

Comment:
Why clone and then iterate mutably? Why not edit in place given that the data structure is not accessed concurrently. (Not particularly important but maybe error prone, when forgetting to write back in certain cases)

Reply:
This is because a few lines down, `get_random_peers` requires an immutable borrow to `self`, so  I couldn't also have a mutable reference to `self.mesh`. 
`get_random_peers` only requires one field of self, so I've now turned it into an associated function which allows the mutable borrow here.

Reply:
I've refactored a few functions to allow for in-place mutation